plugins/default/src/Ide/Plugin/Eval/CodeLens.hs

```diff
@@ -0,0 +1,694 @@
+import           Language.Haskell.LSP.Core      (LspFuncs (getVirtualFileFunc, withIndefiniteProgress),
+                                                 ProgressCancellable (Cancellable))
+import           Language.Haskell.LSP.Types     (ApplyWorkspaceEditParams (ApplyWorkspaceEditParams),
+                                                 CodeLens (CodeLens),
+                                                 CodeLensParams (CodeLensParams, _textDocument),
+                                                 Command (_arguments, _title),
+                                                 ErrorCode (InternalError),
+                                                 List (List), Range (Range),
+                                                 ResponseError (ResponseError),
+                                                 ServerMethod (WorkspaceApplyEdit),
+                                                 TextDocumentIdentifier (..),
+                                                 TextEdit (TextEdit),
+                                                 WorkspaceEdit (WorkspaceEdit),
+                                                 toNormalizedUri)
+import           Language.Haskell.LSP.VFS       (virtualFileText)
+import           Outputable                     (nest, ppr, showSDoc, text,
+                                                 ($$), (<+>))
+import           System.FilePath                (takeFileName)
+import           System.IO                      (hClose)
+import           System.IO.Temp                 (withSystemTempFile)
+import           Text.Read                      (readMaybe)
+import           Util                           (OverridingBool (Never))
+
+-- | Code Lens provider
+-- NOTE: Invoked every time the document is modified, not just when the document is saved.
+codeLens :: CodeLensProvider
+codeLens lsp _state plId CodeLensParams {_textDocument} = response $ do
+  now <- liftIO getCurrentTime
+  let TextDocumentIdentifier uri = _textDocument
+  contents <- liftIO $ getVirtualFileFunc lsp $ toNormalizedUri uri
+  let Just text = virtualFileText <$> contents
+
+  -- Trying to use GerParsedModule to normalise CPP/LHS files, but all comments are removed.
+  -- pr <-
+  --   liftIO $
+  --     runAction "runEvalCmd.GetParsedModule" state $
+  --       use_ GetParsedModule nfp
+  -- liftIO $ dbg "PREPROCESSED !" $ asS . pm_parsed_source $ pr
+
+  -- Use the preprocessor to normalise CPP/LHS files, but we still get an error.
+  -- fp <- handleMaybe "uri" $ uriToFilePath' uri
+  -- let nfp = toNormalizedFilePath' fp
+  -- session :: HscEnvEq <-
+  --   liftIO $
+  --     runAction "runEvalCmd.ghcSession" state $
+  --       use_ GhcSession nfp
+  -- Right (ppContent, dflags) <- liftIO $ evalGhcEnv (hscEnv session) $ runExceptT $ preprocessor fp (Just $ textToStringBuffer content)
+  -- let text = decodeUtf8 $ stringBufferToByteString ppContent
+  -- liftIO $ dbg "PREPROCESSED CONTENT" text
```

Comment:
This should work. What error do you get?

Reply:
The trivial problem was the preprocessor was not even accessible as it is not an exposed module in ghcide.

The main one was that even using the preprocessor I would still get an invalid session (or a similar error) when accessing a CPP file.

Reply:
Make it an exposed module and/or reexport the function in `Development.IDE`, and upstream the change.
Please share the error, I don't think that we should merge this change until CPP works

Reply:
I have submitted a PR to ghcide to export the Preprocessor module.

The latest commit here, adds support for CPP and LHS (Bird-style) for GHC >= 8.8.